### PR TITLE
Fix nested list expansion in JSON-LD

### PR DESCRIPTION
**Pull request by avillar**
# Summary of changes

This PR fixes parsing of deeply nested arrays in JSON-LD:

* by default:
  * Before fix: `[ [ [ 1, 2, 3] , [4, 5] ] ]` --> `( "[1, 2, 3]", "[4, 5]" )` (list contents converted to string literals)
  * After fix: `[ [ [ 1, 2, 3] , [4, 5] ] ]` --> `( 1 2 3 4 5 )` (array is properly flattened)
* when `@container` is set to `@list`:
  * Before fix: `[ [ [ 1, 2, 3] , [4, 5] ] ]` --> `( "[1, 2, 3]", "[4, 5]" )` (list contents converted to string literals)
  * After fix: `[ [ [ 1, 2, 3] , [4, 5] ] ]` --> `( ( ( 1 2 3 ) ( 4 5 ) ) )` (array nesting preserved)

# Checklist

- [X] Checked that there aren't other open pull requests for
  the same change.
- [X] Checked that all tests and type checking passes.
- If the change adds new features or changes the RDFLib public API:
  
  - [ ] Created an issue to discuss the change and get in-principle agreement.
  - [ ] Considered adding an example in `./examples`.
- If the change has a potential impact on users of this project:
  
  - [X] Added or updated tests that fail without the change.
  - [ ] Updated relevant documentation to avoid inaccuracies.
  - [ ] Considered adding additional documentation.
- [X] Considered granting [push permissions to the PR branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork),
  so maintainers can fix minor issues and keep your PR up to date.

### CODE REVIEW

**Comment by aucampia** [comment]
pre-commit.ci autofix

File: test/jsonld/test_nested_arrays.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# -*- coding: UTF-8 -*-
+
+from rdflib import Graph, Literal, URIRef
+from rdflib.collection import Collection
+from rdflib.plugin import Parser, register
+
+register("json-ld", Parser, "rdflib.plugins.parsers.jsonld", "JsonLDParser")
+register("application/ld+json", Parser, "rdflib.plugins.parsers.jsonld", "JsonLDParser")
```
Comment (aucampia) [inline test/jsonld/test_nested_arrays.py:8]: I don't think this is needed, I will check though:

```suggestion
```

Comment (avillar) [reply test/jsonld/test_nested_arrays.py:8]: You may be right (I tried to copy the structure of the other tests)

**Comment by aucampia** [comment]
I'm not really that familiar with JSON-LD as I probably should be, as far as I understand the flattening is pursuant to [JSON-LD 1.1: 9.14 Property Nesting](https://www.w3.org/TR/json-ld11/#property-nesting)

**Review by aucampia — APPROVED** [review]
@avillar thanks for fixing this, and apologies for only reviewing it now. There are some minor comments that I will address with some commits shortly, but the fix looks very good overall.

File: test/jsonld/test_nested_arrays.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        [[1, 2, 3], ["4", 5]],
+        6,
+        [7, { "@id": "_RES_ID_" }]
+    ]
+}
+""".replace(
+    "_PROP_ID_", str(prop)
+).replace(
+    "_RES_ID_", str(res)
+)
```
[the remark this answers is not in the document]
Comment (aucampia) [reply test/jsonld/test_nested_arrays.py]: Pushed changes for this.

File: rdflib/plugins/parsers/jsonld.py [hunk 3]
```diff
@@ -593,3 +597,14 @@ def _to_typed_json_value(value: Any) -> Dict[str, str]:
                 value, separators=(",", ":"), sort_keys=True, ensure_ascii=False
             ),
         }
+
+    @staticmethod
+    def _expand_nested_list(obj_nodes):
```
[the remark this answers is not in the document]
Comment (aucampia) [reply rdflib/plugins/parsers/jsonld.py]: I pushed changes for this now.

File: test/jsonld/test_nested_arrays.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
+# -*- coding: UTF-8 -*-
```
[the remark this answers is not in the document]
Comment (aucampia) [reply test/jsonld/test_nested_arrays.py:1]: I pushed changes for this now.

File: rdflib/plugins/parsers/jsonld.py [hunk 5]
```diff
@@ -593,3 +597,14 @@ def _to_typed_json_value(value: Any) -> Dict[str, str]:
...
+        if not isinstance(obj_nodes, list):
+            result = obj_nodes
+        else:
+            result = [
+                Parser._expand_nested_list(o) if isinstance(o, list) else o
```
[the remark this answers is not in the document]
Comment (aucampia) [reply rdflib/plugins/parsers/jsonld.py]: I pushed changes for this now.

File: rdflib/plugins/parsers/jsonld.py [hunk 6]
```diff
@@ -593,3 +597,14 @@ def _to_typed_json_value(value: Any) -> Dict[str, str]:
                 value, separators=(",", ":"), sort_keys=True, ensure_ascii=False
             ),
         }
+
+    @staticmethod
+    def _expand_nested_list(obj_nodes):
+        if not isinstance(obj_nodes, list):
+            result = obj_nodes
```
[the remark this answers is not in the document]
Comment (aucampia) [reply rdflib/plugins/parsers/jsonld.py]: I pushed changes for this now.

File: rdflib/plugins/parsers/jsonld.py [hunk 7]
```diff
@@ -333,17 +333,21 @@ def _key_to_graph(
...
-                objs = context.get_set(obj)
-                if objs is not None:
-                    obj = objs
-            if isinstance(obj, list):
-                flattened += obj
-                continue
-            flattened.append(obj)
-        obj_nodes = flattened
+        # Flatten deep nested lists
+        def flatten(n):
```
[the remark this answers is not in the document]
Comment (aucampia) [reply rdflib/plugins/parsers/jsonld.py]: I pushed changes for this now.

File: test/jsonld/test_nested_arrays.py (hunk 1 above)
Comment (aucampia) [reply test/jsonld/test_nested_arrays.py]: This is fixed now.

**Comment by aucampia** [comment]
I will merge this before 2023-08-14.

**Comment by avillar** [comment]
Thank you, and thanks for all the fixes!

**Pull request merged**